### PR TITLE
Adds filter and downsample to labjack agent

### PR DESCRIPTION
**Pull request by jlashner**
This adds the ability to filter and downsample labjack data before writing to disk. 

## Description

This adds  a signal processing module with a FIR Filter (used by both this agent and the magpie agent), and a few extra parameters to the labjack agent to enable filtering and downsampling before data is written to disk.

## Motivation and Context

This is mostly to eliminate aliasing in buttkicker accelerometer tests we've been doing on the SAT and LAT. This will allow us to take data at a much higher sample rate, then apply an anti-aliasing filter in software and downsample so that we are writing data at a semi-reasonable cadence to disk. The data rates we're thinking about is raw sampling rates of about 5-10 kHz (for three channels) being filtered then downsampled to ~400 Hz.

## How Has This Been Tested?

Will test this on the SAT this week.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

### CODE REVIEW

File: agents/labjack/labjack_agent.py [hunk 1]
```diff
@@ -309,9 +347,10 @@ def start_acq(self, session, params=None):
 
                 # The labjack outputs at exactly the scan rate but doesn't
                 # generate timestamps. So create them here.
-                timestamps = [cur_time+i/scan_rate for i in range(scans_per_read)]
-                cur_time += scans_per_read/scan_rate
-                data['timestamps'] = timestamps
+                t1 = t0 + read_period
+                timestamps = np.linspace(t0, t1, scans_per_read)
+                data['timestamps'] = timestamps[downsample_mask]
+                t0 = t1
```
Comment (jlashner) [inline agents/labjack/labjack_agent.py:353]: I'm kind of skeptical of this method of generating timestamps, since there's nothing to protect against timestamp drifts if the sample rate isn't exactly what we expect. I didn't want to change the logic here in this PR, but maybe we should think about whether this is the correct way to calculate timestamps

Comment (mhasself) [reply agents/labjack/labjack_agent.py:353]: Agreed this is not good practice... unless the Labjack has access to a realtime clock, then it will drift relative to the agent's computer.  I think it should be fixed for real (unfortunately that's not trivial), but perhaps for now you could at least test for drift and write an error message.

**Comment by jlashner** [comment]
Tested this on SAT and works well now. Here's a comparison of the accelerometer data with and without the aliasing filter, sampled at 8 kHz and downsampled to 400 Hz:
waterfall_unfiltered
waterfall_filtered
Most of the bouncing lines are aliasing artifacts, and adding the anti-aliasing filter both removes most of them and reduces the noise floor.

File: agents/labjack/labjack_agent.py [hunk 2]
```diff
@@ -259,8 +283,32 @@ def start_acq(self, session, params=None):
...
+        downsample_factor = params.get('downsample_factor',
+                                       self.downsample_factor)
+        save_unfiltered = params.get('save_unfiltered', False)
+        chs = params.get('channels', self.chs)
+        if not set(chs).issubset(set(self.chs)):
+            raise ValueError(
+                f"Invalid channels specified! Must be in {self.chs}"
+            )
+        self.log.info(f"Starting taking data for channels {chs}")
+        block_name = '_'.join(sorted(chs))
```
Comment (jlashner) [inline agents/labjack/labjack_agent.py:300]: Question about this: I was having trouble implementing the ability to set the channels being read out through the stream command because the block structure wouldn't match if you stopped the stream, and restarted the stream with a different channel set without restarting the agent. This is the solution I came up with (to make sure the block-name is unique for each combination of channels you could chose. But I just wanted to make sure this wouldn't break anything in the data-loading. I.e.  a field "AIN0V" happens to appear in two different blocks in a single frame, would the hk data loader know how to handle that? Should I change this so every channel is published to a unique block?

Comment (mhasself) [reply agents/labjack/labjack_agent.py:300]: I think that might violate something in the specification (like that a field must not appear in more than one block name for an active provider).  The intended solution for cases like this is that you resign the active provider and start a new provider.  The natural way to do that is to restart the agent.

Alternatively, you could publish to a different feed name (rather than same feed but different block name)... that would trigger a new provider instance (IIRC) **but** then the downstream systems wouldn't naturally group the fields together.

In summary... I don't think this is supported naturally.  A solution might be to add a trigger in the aggregator so it could somehow be told to re-register a provider, so you can change the field contents in a block.

Comment (jlashner) [reply agents/labjack/labjack_agent.py:300]: Ok I think I'll probably just remove this arg. Was useful for testing but not sure how useful it'll be in reality

**Review by BrianJKoopman — COMMENTED** [review]
I don't have a ton to add to this beyond the discussion already ongoing about timestamps and block names. Some small documentation comments for now.

**Comment by jlashner** [comment]
Nope, I think its safe to close this.
